docs/dev/virtual_machine.rst

```diff
@@ -0,0 +1,272 @@
+and integrated with other approaches like ahead of time compilation via a flexible extension
+mechanism.
+
+The virtual machine is designed to strike a balance between performance and flexibility
+when deploying and executing Relay programs, without giving up the benefits of TVM.
+
+Virtual machine (VM) design is a well studied area in programming languages and systems,
+and there have been various virtual machine designs for both full fledged,
+and embedded programing languages.
+Previous language VM designs have been heavily tailored to the execution profile of traditional programs.
+Traditional programs manipulate small scalar values and consist of a large number of low level instructions.
+The sheer quantity of instructions to compute requires instruction execution and dispatch to be extremely efficient.
+In the context of machine learning we manipulate primarily tensor values, using a (relatively)
+low number of high level instructions. ML program's cost centers are expensive operator invocations
+such as GEMM or convolution, over a large input. Due to execution profile exhibited by ML programs
+micro-optimizations present in scalar-VMs are dramatically less important.
+A model’s runtime will  be dominated by executing expensive operators on large inputs.
+
+TVM has provided a strong support for vision models,
+but we want to grow to support a wider variety of models.
+ The graph runtime is able to utilize the fully static nature of the input graphs to perform
+ aggressive optimization such as fully static allocation, and optimal memory reuse.
+ When we introduce models which make use of control-flow, recursion, dynamic shapes, dynamic
+ allocation we must change how execution works.
+
+The rest of this document provides a high level overview of the Relay
+virtual machine design, and its instruction set.
+
+Design
+~~~~~~
+
+The VM's design is focused on simplicity without sacrificing performance.
+In order to accomplish this we have ignored traditional wisdom in scalar
+VM design, and focused on designing a tensor VM.
+
+In the tensor VM setting optimizing for cheap “allocation” of objects (by trying to avoid real allocation,
+reuse of static fragments, and the ability to do dynamic (i.e jagged tensors).
+
+Instruction Set
+~~~~~~~~~~~~~~~
+
+The critical design choice of a VM is the instruction set and their representation.
+The current representation of the instructions is a tagged union, containing the op-code and the data payload.  An important design decision is the level of abstraction of the instructions, and how they take their data, that is RISC vs. CISC and fixed-width instruction encoding vs. variable length. The current version is closer to CISC with complex instructions like AllocTensor, and is variable length due to the inclusion of the shape as part of the instruction. The current instruction set is very high level and corresponds roughly to high level operations in Relay.
+
+### Push
```

Comment:
There's a mismatch between the doc and the implementation. The instructions listed here are still for the stack VM. But the implementation uses register VM.

Reply:
Some of these looks like markdown, please confirm if the doc compiles

Reply:
This is the prototype version I pushed it for Wei and I discuss

Reply:
@jroesch @icemelon9 @tqchen I have updated the doc with register design. Let me check if doc builds.